### ISSUE TITLE
Redshift: Crash in `hasUnprocessedRecords()`

### ISSUE THREAD

**Issue by evantahler** [airbytehq/airbyte#34273]
From Redshift Beta [Sync](https://cloud.airbyte.com/workspaces/502c1fb8-d9ce-4763-99ec-578b3af34ceb/connections/53839d96-2658-4b35-a525-db46a9301a55/job-history#7494491::0)

```
        Caused by: java.lang.NullPointerException: Cannot invoke "io.airbyte.integrations.base.destination.typing_deduping.DestinationHandler$InitialRawTableState.hasUnprocessedRecords()" because the return value of "java.util.Map.get(Object)" is null
                at io.airbyte.integrations.base.destination.typing_deduping.DefaultTyperDeduper.lambda$typeAndDedupe$6(DefaultTyperDeduper.java:274)
                at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:178)
                at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
                at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
```

**Comment by edgao** [comment]
probably two separate things:
* the NPE is fixed, we should just republish dest-redshift with latest CDK? @gisripa solved it in https://github.com/airbytehq/airbyte/pull/33948/files#diff-6544c654b37b2b8e5c945687aff8548bdb979f1455df3597e390d431b0a9ea5a
* there's some very weird exceptions in the logs in general. These two seem most relevant:

```
2024-01-13 05:01:19 destination > com.amazon.redshift.util.RedshiftException: An I/O error occurred while sending to the backend.
2024-01-13 05:01:19 destination >         at com.amazon.redshift.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:402) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeInternal(RedshiftStatementImpl.java:515) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.execute(RedshiftStatementImpl.java:436) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeWithFlags(RedshiftStatementImpl.java:377) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeCachedSql(RedshiftStatementImpl.java:363) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeWithFlags(RedshiftStatementImpl.java:340) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeQuery(RedshiftStatementImpl.java:268) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         at com.amazon.redshift.jdbc.RedshiftDatabaseMetaData.getColumns(RedshiftDatabaseMetaData.java:2095) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         at com.zaxxer.hikari.pool.ProxyDatabaseMetaData.getColumns(ProxyDatabaseMetaData.java:107) ~[HikariCP-5.0.1.jar:?]
2024-01-13 05:01:19 destination >         at com.zaxxer.hikari.pool.HikariProxyDatabaseMetaData.getColumns(HikariProxyDatabaseMetaData.java) ~[HikariCP-5.0.1.jar:?]
2024-01-13 05:01:19 destination >         at io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcDestinationHandler.findExistingTable(JdbcDestinationHandler.java:144) ~[airbyte-cdk-db-destinations-0.11.1.jar:?]
2024-01-13 05:01:19 destination >         at io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcV1V2Migrator.getTableIfExists(JdbcV1V2Migrator.java:56) ~[airbyte-cdk-db-destinations-0.11.1.jar:?]
2024-01-13 05:01:19 destination >         at io.airbyte.integrations.base.destination.typing_deduping.BaseDestinationV1V2Migrator.doesValidV1RawTableExist(BaseDestinationV1V2Migrator.java:130) ~[airbyte-cdk-typing-deduping-0.11.1.jar:?]
2024-01-13 05:01:19 destination >         at io.airbyte.integrations.base.destination.typing_deduping.BaseDestinationV1V2Migrator.shouldMigrate(BaseDestinationV1V2Migrator.java:48) ~[airbyte-cdk-typing-deduping-0.11.1.jar:?]
2024-01-13 05:01:19 destination >         at io.airbyte.integrations.base.destination.typing_deduping.BaseDestinationV1V2Migrator.migrateIfNecessary(BaseDestinationV1V2Migrator.java:27) ~[airbyte-cdk-typing-deduping-0.11.1.jar:?]
2024-01-13 05:01:19 destination >         at io.airbyte.integrations.base.destination.typing_deduping.DefaultTyperDeduper.lambda$prepareTablesFuture$3(DefaultTyperDeduper.java:140) ~[airbyte-cdk-typing-deduping-0.11.1.jar:?]
2024-01-13 05:01:19 destination >         at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768) [?:?]
2024-01-13 05:01:19 destination >         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
2024-01-13 05:01:19 destination >         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
2024-01-13 05:01:19 destination >         at java.base/java.lang.Thread.run(Thread.java:833) [?:?]
2024-01-13 05:01:19 destination > Caused by: java.io.IOException: Unexpected packet type: 0
2024-01-13 05:01:19 destination >         at com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread(QueryExecutorImpl.java:2455) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         at com.amazon.redshift.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1886) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         at com.amazon.redshift.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1878) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         at com.amazon.redshift.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:375) ~[redshift-jdbc42-2.1.0.23.jar:?]
2024-01-13 05:01:19 destination >         ... 19 more
```

```
2024-01-13 05:01:30 destination > com.amazon.redshift.util.RedshiftException: ERROR: Query cancelled on user's request
2024-01-13 05:01:30 destination >         at com.amazon.redshift.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2613)
2024-01-13 05:01:30 destination >         at com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread(QueryExecutorImpl.java:2281)
2024-01-13 05:01:30 destination >         at com.amazon.redshift.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1886)
2024-01-13 05:01:30 destination >         at com.amazon.redshift.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1878)
2024-01-13 05:01:30 destination >         at com.amazon.redshift.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:375)
2024-01-13 05:01:30 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeInternal(RedshiftStatementImpl.java:515)
2024-01-13 05:01:30 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.execute(RedshiftStatementImpl.java:436)
2024-01-13 05:01:30 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeWithFlags(RedshiftStatementImpl.java:377)
2024-01-13 05:01:30 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeCachedSql(RedshiftStatementImpl.java:363)
2024-01-13 05:01:30 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeWithFlags(RedshiftStatementImpl.java:340)
2024-01-13 05:01:30 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeQuery(RedshiftStatementImpl.java:268)
2024-01-13 05:01:30 destination >         at com.amazon.redshift.jdbc.RedshiftDatabaseMetaData.getTables(RedshiftDatabaseMetaData.java:1621)
2024-01-13 05:01:30 destination >         at com.zaxxer.hikari.pool.ProxyDatabaseMetaData.getTables(ProxyDatabaseMetaData.java:67)
2024-01-13 05:01:30 destination >         at com.zaxxer.hikari.pool.HikariProxyDatabaseMetaData.getTables(HikariProxyDatabaseMetaData.java)
2024-01-13 05:01:30 destination >         at io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcDestinationHandler.getInitialRawTableState(JdbcDestinationHandler.java:70)
2024-01-13 05:01:30 destination >         at io.airbyte.integrations.base.destination.typing_deduping.DefaultTyperDeduper.lambda$prepareTablesFuture$3(DefaultTyperDeduper.java:169)
2024-01-13 05:01:30 destination >         at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
2024-01-13 05:01:30 destination >         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
2024-01-13 05:01:30 destination >         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
2024-01-13 05:01:30 destination >         at java.base/java.lang.Thread.run(Thread.java:833)
```

the fact that they're coming from the metadata getTables / getColumns calls is pretty weird though

**Comment by cynthiaxyin** [comment]
this was fixed by #34186
